// ion-bottom-drawer/esm5/modules/ion-bottom-drawer/ion-bottom-drawer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ElementRef, Renderer2, Output, EventEmitter } from '@angular/core';
import { Platform, DomController } from 'ionic-angular';
import { DrawerState } from './drawer-state';
var IonBottomDrawerComponent = /** @class */ (function () {
    function IonBottomDrawerComponent(_element, _renderer, _domCtrl, _platform) {
        this._element = _element;
        this._renderer = _renderer;
        this._domCtrl = _domCtrl;
        this._platform = _platform;
        this.dockedHeight = 50;
        this.shouldBounce = true;
        this.disableDrag = false;
        this.distanceTop = 0;
        this.transition = '0.25s ease-in-out';
        this.state = DrawerState.Bottom;
        this.minimumHeight = 0;
        this.stateChange = new EventEmitter();
        this._BOUNCE_DELTA = 30;
    }
    /**
     * @return {?}
     */
    IonBottomDrawerComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._renderer.setStyle(this._element.nativeElement.querySelector('.ion-bottom-drawer-scrollable-content .scroll-content'), 'touch-action', 'none');
        this._setDrawerState(this.state);
        var /** @type {?} */ hammer = new Hammer(this._element.nativeElement);
        hammer.get('pan').set({ enable: true, direction: Hammer.DIRECTION_VERTICAL });
        hammer.on('pan panstart panend', function (ev) {
            if (_this.disableDrag)
                return;
            switch (ev.type) {
                case 'panstart':
                    _this._handlePanStart();
                    break;
                case 'panend':
                    _this._handlePanEnd(ev);
                    break;
                default:
                    _this._handlePan(ev);
            }
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    IonBottomDrawerComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!changes["state"])
            return;
        this._setDrawerState(changes["state"].currentValue);
    };
    /**
     * @param {?} state
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._setDrawerState = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this._renderer.setStyle(this._element.nativeElement, 'transition', this.transition);
        switch (state) {
            case DrawerState.Bottom:
                this._setTranslateY('calc(100vh - ' + this.minimumHeight + 'px)');
                break;
            case DrawerState.Docked:
                this._setTranslateY((this._platform.height() - this.dockedHeight) + 'px');
                break;
            default:
                this._setTranslateY(this.distanceTop + 'px');
        }
    };
    /**
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handlePanStart = /**
     * @return {?}
     */
    function () {
        this._startPositionTop = this._element.nativeElement.getBoundingClientRect().top;
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handlePanEnd = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (this.shouldBounce && ev.isFinal) {
            this._renderer.setStyle(this._element.nativeElement, 'transition', this.transition);
            switch (this.state) {
                case DrawerState.Docked:
                    this._handleDockedPanEnd(ev);
                    break;
                case DrawerState.Top:
                    this._handleTopPanEnd(ev);
                    break;
                default:
                    this._handleBottomPanEnd(ev);
            }
        }
        this.stateChange.emit(this.state);
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handleTopPanEnd = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (ev.deltaY > this._BOUNCE_DELTA) {
            this.state = DrawerState.Docked;
        }
        else {
            this._setTranslateY(this.distanceTop + 'px');
        }
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handleDockedPanEnd = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        var /** @type {?} */ absDeltaY = Math.abs(ev.deltaY);
        if (absDeltaY > this._BOUNCE_DELTA && ev.deltaY < 0) {
            this.state = DrawerState.Top;
        }
        else if (absDeltaY > this._BOUNCE_DELTA && ev.deltaY > 0) {
            this.state = DrawerState.Bottom;
        }
        else {
            this._setTranslateY((this._platform.height() - this.dockedHeight) + 'px');
        }
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handleBottomPanEnd = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (-ev.deltaY > this._BOUNCE_DELTA) {
            this.state = DrawerState.Docked;
        }
        else {
            this._setTranslateY('calc(100vh - ' + this.minimumHeight + 'px)');
        }
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._handlePan = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        var /** @type {?} */ pointerY = ev.center.y;
        this._renderer.setStyle(this._element.nativeElement, 'transition', 'none');
        if (pointerY > 0 && pointerY < this._platform.height()) {
            if (ev.additionalEvent === 'panup' || ev.additionalEvent === 'pandown') {
                var /** @type {?} */ newTop = this._startPositionTop + ev.deltaY;
                if (newTop >= this.distanceTop)
                    this._setTranslateY(newTop + 'px');
                else if (newTop < this.distanceTop)
                    this._setTranslateY(this.distanceTop + 'px');
                if (newTop > this._platform.height() - this.minimumHeight)
                    this._setTranslateY((this._platform.height() - this.minimumHeight) + 'px');
            }
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    IonBottomDrawerComponent.prototype._setTranslateY = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        this._domCtrl.write(function () {
            _this._renderer.setStyle(_this._element.nativeElement, 'transform', 'translateY(' + value + ')');
        });
    };
    IonBottomDrawerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ion-bottom-drawer',
                    template: "<ion-content class=\"ion-bottom-drawer-scrollable-content\" no-bounce>\n  <ng-content></ng-content>\n</ion-content>\n",
                    styles: [":host{width:100%;height:100%;position:absolute;left:0;z-index:11!important;background-color:#fff;-webkit-transform:translateY(100vh);transform:translateY(100vh)}"]
                },] },
    ];
    /** @nocollapse */
    IonBottomDrawerComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: DomController, },
        { type: Platform, },
    ]; };
    IonBottomDrawerComponent.propDecorators = {
        "dockedHeight": [{ type: Input },],
        "shouldBounce": [{ type: Input },],
        "disableDrag": [{ type: Input },],
        "distanceTop": [{ type: Input },],
        "transition": [{ type: Input },],
        "state": [{ type: Input },],
        "minimumHeight": [{ type: Input },],
        "stateChange": [{ type: Output },],
    };
    return IonBottomDrawerComponent;
}());
export { IonBottomDrawerComponent };
function IonBottomDrawerComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IonBottomDrawerComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IonBottomDrawerComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IonBottomDrawerComponent.propDecorators;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.dockedHeight;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.shouldBounce;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.disableDrag;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.distanceTop;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.transition;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.state;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.minimumHeight;
    /** @type {?} */
    IonBottomDrawerComponent.prototype.stateChange;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._startPositionTop;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._BOUNCE_DELTA;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._element;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._renderer;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._domCtrl;
    /** @type {?} */
    IonBottomDrawerComponent.prototype._platform;
}
